[그리디] 김지우 로또 미션 3,4,5단계 제출합니다.

Ji-Woo-Kim opened the pull request:
kokodak님,
지난 주에 이어 이번 주도 잘 부탁드립니다 🙌 

---

1-2단계 미션 재리뷰 해주신 거랑 제가 공부하면서 반영해보고 싶은 내용을 조합하다 보니 
제 예상보다 코드가 좀 복잡해진 것 같아서 리뷰어님의 의견이 무척 궁금합니다..!

## 고민한 부분 
### - **팩토리 패턴 적용**
지난 미션에서부터 많이 이야기 하기도 했고, 저는 이번 미션에서 로또 생성을 추상화한 전략패턴을 사용했다 보니 의존성 역전 원칙을 신경쓰면서 구현하려고 노렸했습니다.
가장 고민했던 부분은 **하위 모듈이 추상화된 것에 의존하도록 설계하는 것**이었는데요. 
결과적으로 이 어플리케이션에서는 사용자의 입력에 따라 수동 로또가 발급되기 때문에, **이런 패턴들을 적용하는 것이 자연스럽지는 않았던 것 같습니다.** (확장가능성을 고려하면 좋은 설계같지만 저희 미션은 확장가능성이 없으니까요...?)
그럼에도 불구하고, 배운 내용을 직접 적용해 보면 더 확실히 체득할 수 있을 것 같아 일부러 시도해 보았는데, 적절하게 사용한 것인지 궁금합니다...! 🫢 

### - **여전히 `controller`의 역할에 관해... + 도메인 단 로직 분리**
수동 발급 기능이 추가되면서, 지난 주처럼 `Controller`에서 로또를 생성하는 방식이 유지되다 보니 `Controller`가 너무 많은 책임을 지게 되었습니다. 이에 따라 **로또 구매 역할을 담당하는 클래스**(LottoPurchase)를 새로 만들었고, 
기존의 `LottoList`(사용자가 구매한 로또 정보를 관리하던 클래스)를 삭제한 뒤 `LottoResult`를 도입했습니다. 
자연스러운 흐름이라고 생각되지만, 입력값을 domain으로 넘길 때 Controller에서 **데이터 파싱**과 같은 비즈니스 로직을 담당하는 부분이 적절한지 고민이 되었습니다.
결론적으로, 시간이 부족하여 현재 방식대로 유지하기로 했지만, **데이터 파싱을 별도의 클래스로 분리하는 것이 좋을지**에 대한 고민은 여전히 남아 있습니다. 
또, 팩토리 패턴을 적용하면서 클래스 개수가 많아졌기 때문에, 메서드로 해결할 수 있는 부분까지 과도하게 클래스로 분리하는 것이 아닐까 하는 걱정도 듭니다.

위 내용 중점으로 봐주시면 감사하겠습니다 ~! 😄  

---
~~(pr 이후 추가 작성된 부분입니다..)~~
pr 날린 후 깃헙 충돌이 보여서 해결하는 과정에서 약간 문제가 있어서 미션의 연장선으로 계속 보다가
제 어플리케이션이 이상하게 돌아가고 있다는 걸 알아챘습니다...
혹시 리뷰 중에 문제점이 보이신다면 디코로 우선 언급주시면 감사하겠습니다 😭 
그리고 제가 깃헙 충돌을 해결하고 머지한 뒤, 로컬에서 프로젝트를 update 하면
예전 작업 시점으로 돌아갑니다(= 삭제된 파일들이 살아납니다..)
그래서 우선 로컬에서 특정 커밋으로 돌아가서 강제로 푸쉬해 두었는데
리뷰하실 때 문제되지 않으실지 모르겠네요 😭

Ji-Woo-Kim committed:
feat: 로또 1-2단계 구현

Ji-Woo-Kim committed:
Create README.md

Ji-Woo-Kim committed:
refactor: LottoList_불필요한 필드 초기화 삭제 & getter 메서드 불변리스트 반환

Ji-Woo-Kim committed:
Merge remote-tracking branch 'origin/step-1-2' into step-1-2

Ji-Woo-Kim committed:
refactor: 실행 예시와 동일하도록 개행 추가

Ji-Woo-Kim committed:
feat: 로또 당첨 통계 계산 로직을 담은 클래스 (기존 ProfitCalculator에서 분리)

Ji-Woo-Kim committed:
feat: 로또 상금 계산 로직을 담은 클래스 (기존 ProfitCalculator에서 분리)

Ji-Woo-Kim committed:
refactor: 당첨 통계, 상금 계산 로직을 제외한 수익률 계산 로직만을 담도록 개선

Ji-Woo-Kim committed:
refactor: 상수 반영_로또 숫자 범위 경계값 상수 선언

Ji-Woo-Kim committed:
refactor: 변경된 로직에 맞게 인자 수정, 개행 추가

Ji-Woo-Kim committed:
refactor: LottoNumber 객체 반영

Ji-Woo-Kim committed:
refactor: 리뷰 반영(생성기 controller에 주입) & domain 변경에 따른 파라미터 조정

Ji-Woo-Kim committed:
feat: 사용자 입력 값에 대한 검증 로직 구현

Ji-Woo-Kim committed:
test: 로또 번호 검증, 로또 생성, 로또 묶음 생성, 상금 계산 테스트 구현

Ji-Woo-Kim committed:
feat: 보너스볼 입력 메서드 추가

Ji-Woo-Kim committed:
feat: 수동 로또 개수, 번호 입력 메서드 추가

Ji-Woo-Kim committed:
feat: enum 적용_당첨 통계표

Ji-Woo-Kim committed:
feat: 3-4단계 구현

- Factory 패턴 적용
- enum 적용
- LottoList -> LottoPurchase
- 유틸성 메서드명 수정
등...

Ji-Woo-Kim committed:
test: LottoResultTest 수정

Ji-Woo-Kim committed:
test: 코드 로직 변경(수정)

- 로또 결과 부분 테스트 재구성
- 이것저것 사용..

kokodak commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,71 @@
+
+    public void run() {
+        long purchaseAmount = inputView.readPurchaseAmount();
+        int manualCount = inputView.readManualLottoCount();
```
입력된 구입 금액이 수동 로또를 사기에 충분하지 않더라도 프로그램이 돌아가고 있어요! 예외 처리를 해보면 어떨까요?

<img width="246" alt="image" src="https://github.com/user-attachments/assets/a1daa296-7872-4469-a74c-4ac7bcda7d96" />

Ji-Woo-Kim committed:
fix: 보너스볼 당첨 오류 해결

- 보너스볼 당첨되더라도 3,4,5등에 통계되도록 수정
- calculateMatchCount 이동(Lotto.java -> LottoResult.java)
- LottoTest.java 당첨 개수 관련 테스트 메서드 삭제

kokodak commented on src/main/java/domain/generator/LottoGenerator.java:
```diff
@@ -0,0 +1,12 @@
+
+public interface LottoGenerator {
+
+    Lotto generate();
```
이 메서드는 프로덕션 코드에서 큰 의미가 없는 것 같아요. 삭제하는 건 어떨까요?

kokodak commented on src/main/java/factory/LottoGeneratorFactory.java:
```diff
@@ -0,0 +1,20 @@
+
+import java.util.List;
+
+public class LottoGeneratorFactory {
```
> 그럼에도 불구하고, 배운 내용을 직접 적용해 보면 더 확실히 체득할 수 있을 것 같아 일부러 시도해 보았는데, 적절하게 사용한 것인지 궁금합니다...! 🫢

체득하기 위해 시도해보는 것은 좋은 것 같아요. 다만 적절한가? 에 대한 답변은, 저도 자연스럽지는 않은 것 같다고 생각해요.

저는 LottoGenerator가 이름에 Factory가 빠져있을 뿐, 하는 일은 이미 팩토리라고 생각하는데요. 그래서 LottoGeneratorFactory는 의미가 불필요하게 중복된다고 생각합니다.

kokodak commented on src/main/java/domain/WinningLotto.java:
```diff
@@ -0,0 +1,35 @@
+
+public class WinningLotto {
+
+    private final Lotto lotto;
+    private final LottoNumber bonusNumber;
```
오~ WinningLotto를 Lotto 필드를 이용해서 구현해주셨군요! 좋은 접근이라고 생각해요.
지우님이 해주신 방법처럼, 비슷한 유형의 객체를 필드로 가지는 것을 `조합(Composition)`을 사용했다고 표현합니다.

한가지 궁금한 점은, Lotto를 `상속`해서 WinningLotto를 만들어볼 수도 있을 것 같은데, 조합의 방법을 사용하신 이유가 무엇인가요?

Ji-Woo-Kim replied:
미션 진행하면서 제 머릿속에서 `Lotto`는 구매(발급) 시점의 객체, `WinningLotto`는 당첨 통계(결과)를 낼 때 쓰는 객체로 구분되어 있어서, 두 객체의 역할이 다르다고 생각했던 것 같아요. 
구현을 하다 보니 Lotto가 가지고 있는 번호 검증 로직이나 일급 컬렉션 구조가 당첨 번호를 표현하기 좋고 당첨 번호 또한 하나의 로또 번호이기에 결과적으로 WinningLotto 내부에 Lotto를 필드로 사용하게 되었던 것 같아요.
(조합이라는 개념 자체를 인지하고 사용한 것은 아니고 자연스럽게 이런 식으로 구현하게 된 것 같아요. 😅)

그런데 지금 다시 보니 WinningLotto는 Lotto의 구조를 확장해서 기능을 추가하는 형태로도 충분히 바라볼 수 있었을 것 같고, 따라서 상속이 더 좋을 수도 있겠다는 생각이 드네요. 😄 

이 부분은 보완할 수 있도록 고민해보겠습니다!

Ji-Woo-Kim replied:
코드 리팩토링 진행하면서 WinningLotto 부분에서 상속과 조합을 비교해서 계속 고민해봤어요.

상속은 IS-A의 관계, 조합은 HAS-A의 관계라고 생각했을 때,
WinningLotto는 당첨번호와 보너스볼를 모두 갖고 있으므로 
`WinningLotto is a Lotto`가 아니라 `WinningLotto has a Lotto`가 더 맞는 것 같아요.
결과적으로, 다른 변경없이 기존 로직을 그대로 두게 되었습니다!

(처음 리뷰 코멘트를 봤을 땐 상속이 좋은 방향 같아서 긍정적인 반응을 남겼는데 
리팩토링 하다보니 기존 방식(조합)이 맞다는 결론이 나오게 되어 추가 코멘트 남깁니다 😄 )

kokodak commented on src/main/java/domain/LottoPurchase.java:
```diff
@@ -0,0 +1,47 @@
+        this.lottoList = new ArrayList<>();
+        this.lottoList.addAll(manualLottoList); // 수동 로또 추가
+        this.lottoList.addAll(autoLottoList); // 자동 로또 추가
+    }
```
지우님이 어떤 의도를 가지고 LottoPurchase를 만드셨는지는 알겠으나, 너무 과도한 캡슐화가 된 것 같다는 생각을 했어요.

생성자 내부에서 많은 일들이 일어나고 있는데요, 그 때문에 LottoPurchase 객체에서 public하게 열려있는 메서드는 getter 뿐이네요.

LottoPurchase 객체를 사용하는 개발자 입장에서는, 이 객체가 무슨 일을 하는 친구일까 하는 생각이 들 수 있겠어요. 생성자는 따로 네이밍을 할 수 없어서, 내부적으로 어떤 일을 하고 있는지 유추할 수 없기 때문이에요.

어느 정도는 캡슐화를 포기하고, Controller에서 절차적으로 로직의 흐름을 보여주는 것도 좋을 것 같아요.

kokodak commented on src/main/java/domain/Prize.java:
```diff
@@ -0,0 +1,42 @@
+        this.prizeMoney = prizeMoney;
+    }
+
+    public static Prize of(int matchCount, boolean isBonusMatched) {
+        return Arrays.stream(values())
+                .filter(prize -> prize.matchCount == matchCount)
+                .filter(prize -> prize.bonusMatch == isBonusMatched)
+                .findFirst()
+                .orElse(NONE);
+    }
```
저번 스터디 때 배운 정적 팩터리 메서드를 잘 활용해주셨네요~! 💯

kokodak commented on src/main/java/domain/Prize.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.Arrays;
+
+public enum Prize {
+    FIFTH_PRIZE(3, false, 5_000),
+    FOURTH_PRIZE(4, false, 50_000),
+    THIRD_PRIZE(5, false, 1_500_000),
+    SECOND_PRIZE(5, true, 30_000_000),
+    FIRST_PRIZE(6, false, 2_000_000_000),
+    NONE(0, false, 0);
```
Enum을 사용해서 로또 등수에 연관된 데이터들을 묶어주셨군요! 잘 활용해주신 것 같아요. Enum을 사용해보니 어떤 점이 좋으셨나요?

Ji-Woo-Kim replied:
스터디에서 Enum을 이야기할 때 "가독성이 좋다"는 의견에는 솔직히 크게 공감하지 못했었어요.
(상수를 여러 개 나열한 방식도 개인적으로는 읽기 쉬운 것 같다고 생각했어서..ㅎㅎ)
그런데 직접 Enum을 사용해보면서, 단순히 상수를 모아놓는 걸 넘어서 생성자와 관련된 메서드들을 함께 정의해서 하나의 객체처럼 활용할 수 있다는 점에서 Enum의 필요성을 깨닫게 된 것 같아요.!
다만 이번에 처음 사용해보다 보니, 나중에 필요한 부분에 더 효율적으로 활용하려면 연습을 더 해봐야 할 것 같아요 😅

kokodak replied:
말씀주신대로 Enum은 객체처럼 활용할 수 있기 때문에, 메서드 등을 이용해 단순 상수 이상의 행위를 할 수 있어요. 이 미션의 의도는 잘 느끼신 것 같네요 ㅎㅎ 👍

kokodak commented on src/main/java/domain/Prize.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.Arrays;
+
+public enum Prize {
+    FIFTH_PRIZE(3, false, 5_000),
```
```suggestion
public enum Prize {

    FIFTH_PRIZE(3, false, 5_000),
```

사소할 수 있지만, 컨벤션의 통일성은 신경써주는 게 좋다고 생각해요.

kokodak commented on src/main/java/domain/LottoResult.java:
```diff
@@ -0,0 +1,52 @@
+
+        Prize prize = Prize.of(matchCount, isBonusMatched);
+
+        if (prize != Prize.NONE) {
+            matchCountMap.put(prize, matchCountMap.getOrDefault(prize, 0) + 1);
+        }
+
+        if (matchCount >= 3) {
+            Prize basePrize = Prize.of(matchCount, false);
+            matchCountMap.put(basePrize, matchCountMap.getOrDefault(basePrize, 0) + 1);
+        }
```
matchCount가 3 이상일 경우 중복적으로 계산이 되겠네요! 개선이 필요해 보여요.

Ji-Woo-Kim replied:
당첨 개수가 3개, 4개이면서 동시에 보너스볼이 당첨이 되는 경우, 보너스 볼 일치 여부와 관계 없이 당첨 통계에 3개와 4개인 부분에 +1이 되어야 하는데 반영이 안 돼서 추가했던 부분인데요.

당첨개수가 5이고 보너스볼에 당첨된 경우, 2등과 5등이 중복 계산되네요 😅 
수정하도록 하겠습니다!

kokodak commented on src/test/java/domain/LottoTest.java:
```diff
@@ -0,0 +1,97 @@
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("provideInvalidLottoNumbers_Size")
```
오~! `@ParameterizedTest`를 잘 활용해주셨네요! 아주 잘해주셨어요 👍

kokodak reviewed (changes requested):
안녕하세요 지우님~ 👋
다시 만나서 반가워요 ㅎㅎ 전반적으로 코드가 많이 바뀌었다고 느꼈는데, 되게 깔끔해진 것 같아요. 그 사이에 성장해버리신 것 같네요 ㅎㅎ

테스트도 꼼꼼하게 작성하려고 노력하신 게 보였고, 스터디 때 배웠던 내용을 적용해 보려고 하신 점도 매우 좋았습니다.

<img width="541" alt="image" src="https://github.com/user-attachments/assets/9b7cb6eb-3f17-4dcb-ab99-5c33cfbee943" />

도메인에 대한 커버리지도 거의 80%에 근접했네요 👏 

다만 현재 일부 테스트가 깨지고 있는데요, 아마 https://github.com/next-step/java-lotto-clean-playground/pull/102/commits/182ecf90bef06bd4225f72cfd5bbcf370ceec153 커밋 때문인 것 같아요. 무언가 로직을 수정할 때, 항상 테스트를 돌려보면서 의도와 다른 결과가 나오지는 않는지 확인하는 것이 중요해요.

아래에는 지우님이 고민하신 부분에 대한 코멘트를 남겨볼게요.

## 팩토리 패턴 적용

> 결과적으로 이 어플리케이션에서는 사용자의 입력에 따라 수동 로또가 발급되기 때문에, 이런 패턴들을 적용하는 것이 자연스럽지는 않았던 것 같습니다. (확장가능성을 고려하면 좋은 설계같지만 저희 미션은 확장가능성이 없으니까요...?)

일단 위의 남겨주신 코멘트의 의도를 정확히 이해 못했었어요. 구체적으로는,
사용자의 입력에 따라 수동 로또가 발급되기 때문에 —> 이런 패턴들을 적용하는 것이 자연스럽지는 않았던 것 같습니다.
위의 명제 논리 흐름을 잘 파악하지 못했어요. 사용자의 입력에 의존하기 때문에, 전략 패턴/팩토리 패턴을 적용하는 것이 자연스럽지 않았다는 코멘트에 대해 조금 더 설명해주시면 제 의견을 남겨 볼게요.

팩토리 패턴의 사용에 대해서는, 리뷰 코멘트로 제 생각을 남겨 두었습니다.

## 여전히 controller의 역할에 관해... + 도메인 단 로직 분리

> 수동 발급 기능이 추가되면서, 지난 주처럼 Controller에서 로또를 생성하는 방식이 유지되다 보니 Controller가 너무 많은 책임을 지게 되었습니다. 이에 따라 로또 구매 역할을 담당하는 클래스(LottoPurchase)를 새로 만들었고,
기존의 LottoList(사용자가 구매한 로또 정보를 관리하던 클래스)를 삭제한 뒤 LottoResult를 도입했습니다.

1, 2단계 코멘트에도 남겼듯, 저는 Controller가 어느 정도의 도메인 로직에 의존할 수 밖에 없다는 의견에는 변함이 없어요. 지우님도 이 사실을 느끼셨을 것 같은데, 그럼에도 불구하고 LottoPurchase/LottoResult 클래스를 도입해서 도메인 영역을 넓히려는 시도는 매우 좋았다고 생각합니다.

> 자연스러운 흐름이라고 생각되지만, 입력값을 domain으로 넘길 때 Controller에서 데이터 파싱과 같은 비즈니스 로직을 담당하는 부분이 적절한지 고민이 되었습니다.
결론적으로, 시간이 부족하여 현재 방식대로 유지하기로 했지만, 데이터 파싱을 별도의 클래스로 분리하는 것이 좋을지에 대한 고민은 여전히 남아 있습니다.

저도 지우님과 같은 고민을 했던 것 같아요. 참 애매해서 고민했던 지점인데요.
질문 하나 남겨볼게요. 데이터 파싱은 View와 Domain 중 어느 컴포넌트의 비즈니스 로직이라고 생각하시나요?
스스로 고민해보시고, 아래 제 생각을 남길테니 한 번 비교해보며 생각 정리를 해보시면 좋겠어요. (꼭 본인 생각을 정리하고 밑에 읽기!)

위 질문에 대해 저는 이런 생각을 했었어요:
데이터 파싱은 View 컴포넌트에 의존적인 친구라고 생각해요. 콘솔에서 어떻게 입력을 받고, 출력을 하냐에 따라 데이터 파싱의 방법이 변경될 테니까요. 그래서 저는 데이터 파싱에 관련된 클래스를 분리할 거라면, View 패키지 내부에 파싱 클래스를 둘 것 같네요.
만약 클래스를 분리할 정도는 아니라고 한다면, 저는 그냥 View 클래스에서 필요한 처리를 한 뒤에 Domain을 반환시킬 것 같아요. 그게 편리하고 실용적일 것 같아서요.

Ji-Woo-Kim committed:
refactor: Enum 정의 개행 추가

Ji-Woo-Kim committed:
refactor: 로또 생성 로직 이동

- 수동 로또 생성 추상화 (입력값 파싱 후 변환까지 ManualLottoGenerator에서 수행)
- LottoPurchase->LottoList
 : 구매한 로또들의 묶음이라는 의미를 강조하기 위해 rename

Ji-Woo-Kim committed:
refactor: 예외 추가, 오류 수정

- 수동 로또 개수 입력이 구매 가능 개수를 초과한 경우 예외 발생
- 당첨개수 3개 이상인 경우 중복 계산 오류 개선

Ji-Woo-Kim commented:
## 리뷰 반영 정리 🗒️ 

### 구입 금액보다 수동 로또 구매 개수를 많이 입력한 경우 예외 발생
→ 이 예외발생을 어디서 처리할지 고민했습니다. 처음에는 사용자 구매 로또를 발급하는 `LottoList`(전 LottoPurchase)에서 하려고 했는데, `controller`에서 LottoList 생성 시점이 수동 로또 번호를 다 입력 받은 후라 그 이전에 처리해야 겠다는 걸 깨달았어요. 
그래서 `controller`에서 처리하도록 해뒀는데, 이런 경우 로또 한 장 당 가격을 `controller`에서도 또 상수로 선언해둬야 하더라고요… 같은 상수를 여러 클래스에서 선언하는 게 좋은 방법은 아닌 것 같아서 일단은 그냥 1000이라는 숫자로 작성해뒀습니다 🥲
이후 좀 더 고민하다 입력 값에 대한 검증 클래스를 따로 생성하는 방향을 생각했는데, 클래스가 너무 많아질까 싶어 아직 분리하진 않았습니다.
**입력 값 검증(예외 발생 처리 담당)하는 클래스를 따로 분리하는 것에 대해 어떻게 생각하시는 궁금해요!** (현재 제 어플리케이션이 클래스가 너무 많진 않은가요?)

---
### 팩토리 패턴 부분 제거, controller에서 구현체 생성
→ 계속 걸렸던(+ 불편했던) 팩토리 클래스를 제거했습니다.
> 저는 LottoGenerator가 이름에 Factory가 빠져있을 뿐, 하는 일은 이미 팩토리라고 생각하는데요. 그래서 LottoGeneratorFactory는 의미가 불필요하게 중복된다고 생각합니다.

이 말씀에 공감이 많이 되기도 했어요!
그리고 제가 `manualLottoGenerator`를 만들어두고 방치한 채 수동 로또 발급을 따로 로직을 만들어서 하고 있었던 걸 깨닫고 해당 부분 수정했습니다.
결론적으로 자동생성기, 수동생성기를 이용해서 로또를 발급하게 되어 `로또 생성 방식이 2가지가 있구나` 라는게 코드에서 더 잘 보이게 된 것 같아요.
다만 의존성 역전 원칙을 철저히 지키진 못 한 것 같다는 생각이 듭니다.😓

---
### `WinningLotto` **조합** 사용
→ 이 부분은 코멘트 답글로 설명드렸습니다 🙂

---
### `LottoPurchase` → `LottoList`
→ 로또 발급을 담당하던 LottoPurchase의 생성자 내부에서 너무 많은 일이 발생되고 있어 과도한 캡슐화가 된 것 같다는 말씀을 듣고 `controller`에서 발급과 관련된 로직을 일정 부분 나눠 가지도록 했습니다. 이렇게 하다보니 1-2단계 진행했을 때와 비슷하게 controller에 비지니스 로직의 절차가 보이게 되었고, 클래스의 이름도 사용자가 구매한 로또 묶음이라는 의미가 드러나도록 `LottoList`라고 다시 변경하게 되었어요. 결론적으로, 생성 부분은 LottoList의 `create()` 메서드가 담당하게 되었습니다.

---
### 보너스볼 당첨 시 중복 계산되는 부분 오류 수정
→ 조건문 추가해서 중복 계산 방지하도록 수정했습니다!

---

~~수정된 로직에 따라 테스트 코드도 다시 수정하려고 합니다!~~
테스트코드까지 수정했습니다.
생각보다 오래 걸리게 되어 리뷰 반영된 부분을 얼른 공유해드려야 겠다 싶어 미리 코멘트 올려둡니다 🤓 
감사합니당

Ji-Woo-Kim committed:
test: 변경된 domain 로직 적용

Ji-Woo-Kim committed:
refactor

kokodak commented on src/main/java/controller/LottoController.java:
```diff
@@ -1,43 +1,64 @@
     public void run() {
-        final long purchaseAmount = InputView.readPurchaseAmount();
-        final LottoList lottoList = new LottoList(purchaseAmount, generator);
+        long purchaseAmount = inputView.readPurchaseAmount();
+        final int manualLottoCount = inputView.readManualLottoCount();
+        if (purchaseAmount/1000 < manualLottoCount) {
+            throw new IllegalArgumentException("수동 로또 개수가 구매 가능한 로또 개수를 초과했습니다.");
+        }
```
👍 💯

kokodak reviewed (approved):
LGTM :)
